8298728: Cells in VirtualFlow jump after resizing

johanvos opened the pull request:
When recalculating sizes, we often don't want the current index and/or offset to change.

Allow to fix the index/offset when doing recalculations.

Fix JDK-8298728

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8298728](https://bugs.openjdk.org/browse/JDK-8298728): Cells in VirtualFlow jump after resizing


### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/974/head:pull/974` \
`$ git checkout pull/974`

Update a local copy of the PR: \
`$ git checkout pull/974` \
`$ git pull https://git.openjdk.org/jfx pull/974/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 974`

View PR using the GUI difftool: \
`$ git pr show -t 974`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/974.diff">https://git.openjdk.org/jfx/pull/974.diff</a>

</details>

andy-goryachev-oracle commented:
I am trying to understand what the behavioral difference is between this branch and the current master.  The ticket JDK-8298728 does not seem to specify the exact scenario, so I tried the following code with the large model (3000 lines) with a wrapping Text as graphic node.  Resizing both vertically and horizontally produces what I this is identical result where the top visible selected cell scrolls out of the view in both cases.

cell factory:
```
                       TableColumn<String,String> c = new TableColumn<>();
                        table.getColumns().add(c);
                        c.setText("C" + table.getColumns().size());
                        c.setCellValueFactory((f) -> new SimpleStringProperty(describe(c)));
                        c.setCellFactory((r) -> {
                            return new TableCell<>() {
                                @Override
                                protected void updateItem(String item, boolean empty) {
                                    super.updateItem(item, empty);
                                    Text t = new Text("11111111111111111111111111111111111111111111111111111111111111111111111111111111111111\n2\n3\n");
                                    t.wrappingWidthProperty().bind(widthProperty());
                                    setPrefHeight(USE_COMPUTED_SIZE);
                                    setGraphic(t);
                                }
                            };
                        });

```

I can attache the whole program to the jira ticket.

johanvos commented:
> I am trying to understand what the behavioral difference is between this branch and the current master. 

I added a new test in VirtualFlowTest. It fails without the patch, and succeeds with the patch. The change is that if a cell is shown with some given content, then a cell is resized, and then a scroll operation is performed, the result of this scroll could be unexpected. I added comments in the added test so that the flow should be clear. Does this help?

andy-goryachev-oracle commented:
I think this might be the case when unit tests are insufficient.  
With this change, I notice something else: select a cell somewhere in the middle of the table, click on the scroll bar to page up, then click on the other side of the scroll bar to page down.  The existing code works correctly, the new code does not - the position is different, sometimes the selected cell is scrolled outside of the view.
see "does not page back to the same spot.mp4" in jira

kevinrushforth commented:
/reviewers 2

johanvos commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:
```diff
@@ -2313,23 +2315,29 @@ void setCellDirty(int index) {
+        int oldIndex = computeCurrentIndex();
+        double cellLength = getOrCreateCellSize(index);
+        if (index > 0) getOrCreateCellSize(index - 1);
+        if (index < getCellCount() -1) getOrCreateCellSize(index + 1);
```
True -- I'll change that.

andy-goryachev-oracle commented:
1. I see the same issue with page up/page down described earlier (see the video clip attached to the jira ticket https://bugs.openjdk.org/secure/attachment/102048/does%20not%20page%20back%20to%20the%20same%20spot.mov )

2. I am still not clear what exactly the problem is.  The jira ticket contains conditionals ("if we want..."), rather than a typical `steps to reproduce -> expected to see -> actual behavior` steps.

Could you comment on #1 and #2 please?

FlorianKirmaier commented:
I can confirm that this fixes every remaining issue related to ListView to my knowledge.

My test application (which was an extended version of the test in this ticket: https://bugs.openjdk.org/browse/JDK-8277785) works now 100% correctly.

The application at Zeiss works now without any ListView-related workaround.

So, on my side - with this fix, all ListView issues known to me are fixed.

johanvos commented:
> 1. I see the same issue with page up/page down described earlier (see the video clip attached to the jira ticket https://bugs.openjdk.org/secure/attachment/102048/does%20not%20page%20back%20to%20the%20same%20spot.mov )
> 2. I am still not clear what exactly the problem is.  The jira ticket contains conditionals ("if we want..."), rather than a typical `steps to reproduce -> expected to see -> actual behavior` steps.
> 
> Could you comment on #1 and #2 please?

#1: I had a look at that video and I don't think this is a bug. Nowhere in the spec or javadoc it is mentioned that a page-up followed by a page-down should give the same position. This would imho actually be a bad requirement, as in case sizes are changing (which happens to be the case in most real-world scenario's I've seen so far) this would conflict with other requirements. One might argue that "in case the size of individual cells didn't change", the scroll-up/scroll-down should result in a same "offset" as before the operation but that is not a requirement so far. Could be discussed in another issue.

#2. I don't see what is missing. It seems to me both the ticket and the new test in the PR show the scenario of what goes wrong before the patch, and what is expected (in the assert). I can copy it here again, but I'm not understanding your question I think.

andy-goryachev-oracle commented:
Thank you for commenting, Johan.

Interesting.  Why would sizes change in #1 scenario, since there are no changes in the component width and the scroll bar state?  But I might disagree - it's the most natural case when pgup/pgdn results in the same state, as any text editor would demonstrate.  In fact, I have a similar problem with "If we want the position of the thumb constant," - I'd rather see the thumb reflecting reality as much as it could, so it is ok for it to jump, as long as the **selected cell** remains in the same position on screen.  It looks like the ticket talks about the "first cell" instead which would be an acceptable replacement.

As for #2 - it might be just me, I cannot translate these words into a scenario.  Are you saying that, for example, resizing and immediately issuing pgup/pgdn (via click on a scrollbar) would shift the viewport too far/not enough?  Or is there another scenario?  What are the exact steps that result in erroneous behavior, in human words?

aghaisas reviewed (commented):
Overall fix looks good to me.

I have identified two very minor typos.

johanvos commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/VirtualFlowTest.java:
```diff
@@ -1424,6 +1424,73 @@ protected double computeMaxHeight(double width) {
+        assertEquals(300, thirdCell.getLayoutY(), 1.);
+
+
+        for (int i =0 ; i < heights.length; i++) {
```
fixed

johanvos commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:
```diff
@@ -2313,23 +2315,33 @@ void setCellDirty(int index) {
+        if (index > 0) {
+            getOrCreateCellSize(index - 1);
+        }
+        if (index < getCellCount() -1) {
```
fixed

andy-goryachev-oracle reviewed (approved):
lgtm
